driver-core/src/main/com/mongodb/internal/session/ServerSessionPool.java

```diff
@@ -77,60 +72,40 @@ public ServerSessionPool(final Cluster cluster, @Nullable final ServerApi server
 
     public ServerSession get() {
         isTrue("server session pool is open", !closed);
-        ServerSessionImpl serverSession = serverSessionPool.get();
-        while (shouldPrune(serverSession)) {
-            serverSessionPool.release(serverSession, true);
-            serverSession = serverSessionPool.get();
+        ServerSessionImpl serverSession = available.pollLast();
+        while (serverSession != null && shouldPrune(serverSession)) {
+            serverSession.close();
+            serverSession = available.pollLast();
+        }
+        if (serverSession == null) {
+            serverSession = new ServerSessionImpl();
         }
+        inUseCount.increment();
         return serverSession;
     }
 
     public void release(final ServerSession serverSession) {
-        serverSessionPool.release((ServerSessionImpl) serverSession);
-        serverSessionPool.prune();
-    }
-
-    public void close() {
-        try {
-            closing = true;
-            serverSessionPool.close();
-
-            List<BsonDocument> identifiers;
-            synchronized (this) {
-                identifiers = new ArrayList<>(closedSessionIdentifiers);
-                closedSessionIdentifiers.clear();
-            }
-            endClosedSessions(identifiers);
-        } finally {
-            closed = true;
+        inUseCount.decrement();
+        ServerSessionImpl serverSessionImpl = (ServerSessionImpl) serverSession;
+        if (serverSessionImpl.isMarkedDirty()) {
+            serverSessionImpl.close();
+        } else {
+            available.addLast(serverSessionImpl);
         }
     }
 
-    public int getInUseCount() {
-        return serverSessionPool.getInUseCount();
+    public long getInUseCount() {
+        return inUseCount.sum();
     }
 
-    private void closeSession(final ServerSessionImpl serverSession) {
-        serverSession.close();
-        // only track closed sessions when pool is in the process of closing
-        if (!closing) {
-            return;
-        }
-
-        List<BsonDocument> identifiers = null;
-        synchronized (this) {
-            closedSessionIdentifiers.add(serverSession.getIdentifier());
-            if (closedSessionIdentifiers.size() == END_SESSIONS_BATCH_SIZE) {
-                identifiers = new ArrayList<>(closedSessionIdentifiers);
-                closedSessionIdentifiers.clear();
-            }
-        }
-        if (identifiers != null) {
-            endClosedSessions(identifiers);
-        }
+    public void close() {
+        closed = true;
+        endClosedSessions();
     }
 
-    private void endClosedSessions(final List<BsonDocument> identifiers) {
+    private void endClosedSessions() {
+        List<BsonDocument> identifiers = drainPool();
+        available.clear();
```

Comment:
Given that we have just drained `available`, we don't need to clear it.

Reply:
Done